Add crud support for BPF_MAP_TYPE_BLOOM_FILTER

barthr opened the pull request:
This PR fixes issue: #777

danielocfb commented:
Seems fine to me, thanks! Would you be able to add a test for this, by any chance? Also, can you please squash commits into a single one?

barthr commented:
Yes will do!

barthr commented:
So couple of things, when I started working on the tests I noticed that there is no way to do a proper lookup using the current lookup functionality. So I've added another function which should be used when u are using a bloom filter. I made it conform the spec as per (https://docs.kernel.org/bpf/map_bloom_filter.html). I also added some small refactoring. The PR turned out a little bit bigger than expected. Let me know what you think, I've only recently (this week) started to learn rust so it's all quite new to me.

danielocfb commented on libbpf-rs/src/map.rs:
```diff
@@ -631,7 +632,14 @@ impl MapHandle {
+    /// If the map is of type bloom_filter the function [`MapHandle::lookup_bloom_filter()`] must be used
     pub fn lookup(&self, key: &[u8], flags: MapFlags) -> Result<Option<Vec<u8>>> {
+        if self.map_type().is_bloom_filter() {
+            return Err(Error::with_invalid_data(format!(
+                "lookup_bloom_filter() must be used for bloom_filter maps (type of the map is {})",
+                self.map_type(),
```
```suggestion
            return Err(Error::with_invalid_data("lookup_bloom_filter() must be used for bloom filter maps"
```

danielocfb reviewed (approved):
I am unsure how exactly I feel about `lookup_bloom_filter`, but I suppose there is precedent, I don't see a better way, and the alternative of using `lookup` seems too....dodgy. So why not. Thanks!
Can you please take a look at the one suggestion I left?

barthr commented:
I get your view. I tried looking into staying as close as possible to existing calls and existing API's however it would make the lookup code a bit ugly and violating the principle of least suprises. 

So that is why I opted for the specific method for the bloom filter map. However, there might also be solutions which use a more polymorphism approach. Like a trait with a default lookup method with custom overrides for certain map types. I think for this PR this is out of scope.

Unfortunately, the libbpf api is a bit funky since the same methods mean different things on different types, that is the downside of making everything a map I suppose.

I will look at your comments tomorrow.
